In this project we will clean the data and analyze the included car listings. We will import the Numpy and Pandas libraries for assistance.
We start by reading in the CSV file, and having a high level look via df.info() and df.head().

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [4]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [5]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.rename({"yearOfRegistration":"registration_year"}, axis=1, inplace=True)
autos.rename({"monthOfRegistration":"registration_month"}, axis=1, inplace=True)
autos.rename({"notRepairedDamage":"unrepaired_damage"}, axis=1, inplace=True)
autos.rename({"dateCreated":"ad_created"}, axis=1, inplace=True)
autos.rename({"dateCrawled":"date_crawled"}, axis=1, inplace=True)
autos.rename({"offerType":"offer_type"}, axis=1, inplace=True)
autos.rename({"vehicleType":"vehicle_type"}, axis=1, inplace=True)
autos.rename({"powerPS":"power_ps"}, axis=1, inplace=True)
autos.rename({"fuelType":"fuel_type"}, axis=1, inplace=True)
autos.rename({"nrOfPictures":"nr_of_pictures"}, axis=1, inplace=True)
autos.rename({"postalCode":"postal_code"}, axis=1, inplace=True)
autos.rename({"lastSeen":"last_seen"}, axis=1, inplace=True)

In [8]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [9]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

I made the column headers more descriptive by making them snakecase.

In [11]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [12]:
print(autos["registration_year"].value_counts())
print(autos["price"].value_counts())
print(autos["odometer"].value_counts())

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
2006    2708
2001    2703
2002    2533
1998    2453
2007    2304
2008    2231
2009    2098
1997    2028
2011    1634
2010    1597
2017    1453
1996    1444
2012    1323
2016    1316
1995    1313
2013     806
2014     666
1994     660
2018     492
1993     445
2015     399
1990     395
1992     391
1991     356
1989     181
        ... 
1950       3
1955       2
9000       2
1954       2
1800       2
1957       2
1941       2
1951       2
1934       2
4100       1
4800       1
1953       1
1111       1
1927       1
6200       1
4500       1
1943       1
5911       1
1939       1
1938       1
2800       1
8888       1
1000       1
1500       1
1948       1
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64
$0          1421
$500         781
$1,500       734
$2,500       643
$1,200       639
$1,000       639
$600         531
$800         498
$3,500       498
$2,0

The columns with mostly one value may be considered for deletion. e.g seller, offer type, etc
Number of pictures column has to be checked as the data impliens there isn't a single picture on the database.
Registration year needs a closer look as some years are too in the future and too in the past.

Also, price and odometer columns are numeric data stored as text. These will need to be changed.

In [13]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)
autos["price"] = autos["price"].round(3)
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(int)
autos["odometer"] = autos["odometer"].round(5)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [15]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

In [16]:
autos["odometer_km"].unique().shape

(13,)

In [17]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [18]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [19]:
autos["odometer_km"].value_counts().head().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [20]:
autos["price"].sort_values(ascending=False).head()

39705    99999999
42221    27322222
39377    12345678
47598    12345678
27371    12345678
Name: price, dtype: int64

In [21]:
autos.loc[42221,:]

date_crawled          2016-03-08 20:39:05
name                    Leasinguebernahme
seller                             privat
offer_type                        Angebot
price                            27322222
abtest                            control
vehicle_type                    limousine
registration_year                    2014
gearbox                           manuell
power_ps                              163
model                                  c4
odometer_km                         40000
registration_month                      2
fuel_type                          diesel
brand                             citroen
unrepaired_damage                     NaN
ad_created            2016-03-08 00:00:00
nr_of_pictures                          0
postal_code                         76532
last_seen             2016-03-08 20:39:05
Name: 42221, dtype: object

In [22]:
autos["price"].value_counts().tail(100).sort_index(ascending = True)

14        1
173       1
188       1
205       1
238       1
269       1
686       1
846       1
860       1
910       1
1039      1
1070      1
1295      1
1325      1
1485      1
1645      1
1659      1
1775      1
1933      1
1935      1
2095      1
2479      1
2589      1
2671      1
2895      1
3020      1
3055      1
3119      1
3279      1
3660      1
         ..
21959     1
22790     1
23590     1
24550     1
25698     1
25699     1
27299     1
27650     1
28290     1
28850     1
29600     1
29699     1
30899     1
33980     1
41800     1
41850     1
46200     1
47499     1
48600     1
48888     1
56800     1
68300     1
68750     1
69900     1
69999     1
86500     1
115991    1
151990    1
175000    1
194000    1
Name: price, Length: 100, dtype: int64

In [23]:
autos["price"].unique().shape

(2357,)

In [24]:
autos["price"].max()

99999999

In [25]:
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

We will now remove outliers from the two columns

In [26]:
autos = autos[autos["price"].between(1000,5000000)]

In [27]:
autos = autos[autos["odometer_km"].between(1000,150000)]

In [28]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  3.863200e+04       38632.000000  38632.000000   38632.000000   
mean   7.498207e+03        2005.678194    128.928531  122775.807621   
std    2.535787e+04          86.678597    215.903018   40800.492791   
min    1.000000e+03        1000.000000      0.000000    5000.000000   
25%    2.200000e+03        2001.000000     80.000000  100000.000000   
50%    4.350000e+03        2005.000000    116.000000  150000.000000   
75%    8.950000e+03        2009.000000    160.000000  150000.000000   
max    3.890000e+06        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        38632.000000         38632.0  38632.000000  
mean             5.998835             0.0  51690.564765  
std              3.591214             0.0  25689.463891  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  31224.000000  
50%              6.000000             0.0  50858.000000  
75%              9.000000             0.0  72534.000000  
max             12.000000             0.0  99998.000000

In [29]:
autos["odometer_km"].value_counts()

150000    23317
125000     4341
100000     1860
90000      1569
80000      1334
70000      1154
60000      1099
50000       987
40000       795
30000       748
20000       692
5000        508
10000       228
Name: odometer_km, dtype: int64

In [30]:
len(autos["odometer_km"])

38632

In [31]:
print(autos['date_crawled'].str[:10])
print("\n")
print(autos['ad_created'].str[:10])
print("\n")
print(autos['last_seen'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
7        2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
17       2016-03-29
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
28       2016-03-19
29       2016-04-02
31       2016-03-14
33       2016-03-15
34       2016-04-05
35       2016-03-29
36       2016-03-16
37       2016-03-23
38       2016-03-21
39       2016-03-11
40       2016-03-07
41       2016-03-10
            ...    
49962    2016-03-14
49963    2016-03-26
49964    2016-03-10
49965    2016-03-11
49966    2016-04-02
49967    2016-03-12
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49979    2016-03-20
49981    2016-03-15
49982    2016-03-29
49985    2016-04-02
49986    2016-04-04


In [32]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025549
2016-03-06    0.013875
2016-03-07    0.035126
2016-03-08    0.032615
2016-03-09    0.032460
2016-03-10    0.033314
2016-03-11    0.032797
2016-03-12    0.037378
2016-03-13    0.015997
2016-03-14    0.036628
2016-03-15    0.033625
2016-03-16    0.029069
2016-03-17    0.030493
2016-03-18    0.012839
2016-03-19    0.035126
2016-03-20    0.038155
2016-03-21    0.037301
2016-03-22    0.032538
2016-03-23    0.032201
2016-03-24    0.029017
2016-03-25    0.030519
2016-03-26    0.033107
2016-03-27    0.031399
2016-03-28    0.035359
2016-03-29    0.033987
2016-03-30    0.033055
2016-03-31    0.031399
2016-04-01    0.034609
2016-04-02    0.036291
2016-04-03    0.039139
2016-04-04    0.036912
2016-04-05    0.013357
2016-04-06    0.003262
2016-04-07    0.001501
Name: date_crawled, dtype: float64

**The date crawled dates are between March and April. Date crawled seems to have a wide distribution with no particular date concentration**

In [33]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
2016-01-03    0.000026
2016-01-07    0.000026
2016-01-10    0.000052
2016-01-13    0.000026
2016-01-14    0.000026
2016-01-16    0.000026
2016-01-22    0.000026
2016-01-27    0.000078
2016-01-29    0.000026
2016-02-01    0.000026
2016-02-02    0.000052
2016-02-05    0.000052
2016-02-07    0.000026
2016-02-09    0.000026
2016-02-11    0.000026
2016-02-12    0.000052
2016-02-14    0.000052
2016-02-16    0.000026
2016-02-17    0.000026
2016-02-18    0.000052
2016-02-19    0.000078
2016-02-20    0.000026
2016-02-21    0.000052
2016-02-22    0.000026
2016-02-23    0.000104
                ...   
2016-03-09    0.032641
2016-03-10    0.032978
2016-03-11    0.033055
2016-03-12    0.037119
2016-03-13    0.017654
2016-03-14    0.034971
2016-03-15    0.033444
2016-03-16    0.029639
2016-03-17    0.030182
2016-03-18    0.013279
2016-03-19    0.034065
2016-03-20    0.038258
2016-03-21 

**There is a greater frequency of ads creation in 2016 than in 2015**

In [34]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001087
2016-03-06    0.003572
2016-03-07    0.004556
2016-03-08    0.006238
2016-03-09    0.008905
2016-03-10    0.009811
2016-03-11    0.011726
2016-03-12    0.022184
2016-03-13    0.008387
2016-03-14    0.011985
2016-03-15    0.014988
2016-03-16    0.015454
2016-03-17    0.026377
2016-03-18    0.007377
2016-03-19    0.014599
2016-03-20    0.019802
2016-03-21    0.019673
2016-03-22    0.020786
2016-03-23    0.017913
2016-03-24    0.018534
2016-03-25    0.017757
2016-03-26    0.016075
2016-03-27    0.014082
2016-03-28    0.019440
2016-03-29    0.020786
2016-03-30    0.023452
2016-03-31    0.022727
2016-04-01    0.023193
2016-04-02    0.024928
2016-04-03    0.024436
2016-04-04    0.023374
2016-04-05    0.131161
2016-04-06    0.234676
2016-04-07    0.139962
Name: last_seen, dtype: float64

**There is a date concentration in April 2016. 49% of last seen were on three days in April**

In [35]:
autos["registration_year"].describe()

count    38632.000000
mean      2005.678194
std         86.678597
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

There are outliers like year 1000 and year 9999 which are too old and too fururistic to be real. We remove these outliers.

In [36]:
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       3
1959       6
1960      17
1961       6
1962       4
1963       8
1964       9
1965      17
1966      21
1967      25
        ... 
1999    1725
2000    2001
2001    2065
2002    2135
2003    2477
2004    2608
2005    2785
2006    2652
2007    2259
2008    2204
2009    2077
2010    1584
2011    1617
2012    1306
2013     795
2014     649
2015     361
2016     651
2017    1008
2018     398
2019       1
2800       1
4100       1
4500       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       2
Name: registration_year, Length: 91, dtype: int64

The earliest acceptable year is obviously 1927 as anything before that is medieval. We will remove any row with registration year after 2016, which is the last year any ad was seen.

In [37]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [38]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  3.721200e+04       37212.000000  37212.000000   37212.000000   
mean   7.605764e+03        2003.967618    129.496856  122540.175212   
std    2.581081e+04           7.021028    197.414831   40886.472839   
min    1.000000e+03        1927.000000      0.000000    5000.000000   
25%    2.222000e+03        2001.000000     82.000000  100000.000000   
50%    4.450000e+03        2005.000000    117.000000  150000.000000   
75%    9.000000e+03        2008.000000    160.000000  150000.000000   
max    3.890000e+06        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        37212.000000         37212.0  37212.000000  
mean             6.042755             0.0  51828.871386  
std              3.574749             0.0  25680.166234  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  31275.000000  
50%              6.000000             0.0  51065.000000  
75%              9.000000             0.0  72632.250000  
max             12.000000             0.0  99998.000000

In [39]:
autos["registration_year"].value_counts(normalize=True)

2005    0.074841
2006    0.071267
2004    0.070085
2003    0.066565
2007    0.060706
2008    0.059228
2002    0.057374
2009    0.055815
2001    0.055493
2000    0.053773
1999    0.046356
2011    0.043454
2010    0.042567
2012    0.035096
1998    0.034478
2013    0.021364
1997    0.021149
2016    0.017494
2014    0.017441
1996    0.014565
1995    0.011905
2015    0.009701
1994    0.007283
1992    0.005885
1993    0.005858
1991    0.005563
1990    0.005079
1989    0.003305
1988    0.002849
1985    0.002150
          ...   
1976    0.000564
1977    0.000537
1975    0.000484
1969    0.000484
1965    0.000457
1960    0.000457
1964    0.000242
1963    0.000215
1959    0.000161
1961    0.000161
1956    0.000107
1962    0.000107
1937    0.000107
1958    0.000081
1954    0.000054
1955    0.000054
1941    0.000054
1957    0.000054
1934    0.000054
1951    0.000054
1953    0.000027
1943    0.000027
1927    0.000027
1929    0.000027
1931    0.000027
1950    0.000027
1948    0.000027
1938    0.0000

Over 1000 entries (reg year after 2016) have been deleted. Most of the cars were registered in the early 2000s

In [40]:
autos["brand"].value_counts()

volkswagen        7845
bmw               4665
mercedes_benz     4152
audi              3631
opel              3314
ford              2185
renault           1387
peugeot           1038
fiat               784
skoda              709
seat               643
smart              618
toyota             544
mazda              530
citroen            517
nissan             507
mini               405
hyundai            400
sonstige_autos     391
volvo              334
kia                286
porsche            278
honda              273
mitsubishi         256
chevrolet          246
alfa_romeo         232
suzuki             213
dacia              122
chrysler           118
jeep               103
land_rover          98
jaguar              69
subaru              64
daihatsu            63
saab                51
daewoo              34
trabant             32
rover               27
lancia              25
lada                23
Name: brand, dtype: int64

In [41]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'audi', 'renault', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'jeep', 'saab',
       'volvo', 'nissan', 'jaguar', 'skoda', 'subaru', 'fiat',
       'mitsubishi', 'chevrolet', 'hyundai', 'honda', 'kia', 'citroen',
       'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'rover',
       'daihatsu', 'daewoo', 'lancia', 'lada'], dtype=object)

In [42]:
top_car_count = autos["brand"].value_counts().head(20).index

I have chosen to aggregate on the top 20 most popular brands. The working is below.

In [43]:
top_20_cars_price = {}
for car in top_car_count:
    selected_rows = autos[autos["brand"] == car]
    mean_price = selected_rows["price"].mean()
    top_20_cars_price[car] = mean_price

In [44]:
print(top_20_cars_price)

{'smart': 3780.4692556634304, 'renault': 3590.942321557318, 'citroen': 4614.970986460348, 'peugeot': 3955.169556840077, 'nissan': 6428.428007889546, 'sonstige_autos': 27654.273657289003, 'mercedes_benz': 9302.614402697494, 'fiat': 4008.174744897959, 'skoda': 6836.696755994359, 'volvo': 6151.583832335329, 'audi': 10322.269347287249, 'ford': 5786.703432494279, 'volkswagen': 6898.376545570427, 'bmw': 9381.89174705252, 'hyundai': 6181.8725, 'toyota': 5573.57169117647, 'mazda': 5309.526415094339, 'seat': 5638.640746500778, 'opel': 4219.954737477368, 'mini': 10715.237037037037}


In [45]:
top_20_cars_miles = {}
for car in top_car_count:
    selected_rows = autos[autos["brand"] == car]
    mean_mileage = selected_rows["odometer_km"].mean()
    top_20_cars_miles[car] = mean_mileage

In [46]:
print(top_20_cars_miles)

{'smart': 97775.08090614887, 'renault': 121423.93655371305, 'citroen': 114245.64796905222, 'peugeot': 122341.04046242774, 'nissan': 110335.3057199211, 'sonstige_autos': 89411.76470588235, 'mercedes_benz': 130062.6204238921, 'fiat': 107901.78571428571, 'skoda': 110063.46967559944, 'volvo': 137230.53892215568, 'audi': 127491.04929771413, 'ford': 119622.42562929062, 'volkswagen': 125771.82919056724, 'bmw': 132005.359056806, 'hyundai': 101862.5, 'toyota': 113740.80882352941, 'mazda': 119981.1320754717, 'seat': 116104.19906687403, 'opel': 123952.92697646348, 'mini': 88679.01234567902}


In [47]:
bmp_series = pd.Series(top_20_cars_price)
print(bmp_series)

audi              10322.269347
bmw                9381.891747
citroen            4614.970986
fiat               4008.174745
ford               5786.703432
hyundai            6181.872500
mazda              5309.526415
mercedes_benz      9302.614403
mini              10715.237037
nissan             6428.428008
opel               4219.954737
peugeot            3955.169557
renault            3590.942322
seat               5638.640747
skoda              6836.696756
smart              3780.469256
sonstige_autos    27654.273657
toyota             5573.571691
volkswagen         6898.376546
volvo              6151.583832
dtype: float64


In [48]:
top_mileage_series = pd.Series(top_20_cars_miles)
print(top_mileage_series)

audi              127491.049298
bmw               132005.359057
citroen           114245.647969
fiat              107901.785714
ford              119622.425629
hyundai           101862.500000
mazda             119981.132075
mercedes_benz     130062.620424
mini               88679.012346
nissan            110335.305720
opel              123952.926976
peugeot           122341.040462
renault           121423.936554
seat              116104.199067
skoda             110063.469676
smart              97775.080906
sonstige_autos     89411.764706
toyota            113740.808824
volkswagen        125771.829191
volvo             137230.538922
dtype: float64


In [49]:
top_20_cars = pd.DataFrame(top_mileage_series, columns=['mean_mileage'])
top_20_cars["mean_price"] = pd.DataFrame(bmp_series, columns=['mean_price'])

top_20_cars.sort_values("mean_price", ascending = False)

mean_mileage    mean_price
sonstige_autos   89411.764706  27654.273657
mini             88679.012346  10715.237037
audi            127491.049298  10322.269347
bmw             132005.359057   9381.891747
mercedes_benz   130062.620424   9302.614403
volkswagen      125771.829191   6898.376546
skoda           110063.469676   6836.696756
nissan          110335.305720   6428.428008
hyundai         101862.500000   6181.872500
volvo           137230.538922   6151.583832
ford            119622.425629   5786.703432
seat            116104.199067   5638.640747
toyota          113740.808824   5573.571691
mazda           119981.132075   5309.526415
citroen         114245.647969   4614.970986
opel            123952.926976   4219.954737
fiat            107901.785714   4008.174745
peugeot         122341.040462   3955.169557
smart            97775.080906   3780.469256
renault         121423.936554   3590.942322

There seems to be a slight inverse proportional relationship between mileage and price. Sonstige autos have the highest mean price and the lowest mean mileage, while it is vice versa for volvo.